src/xref2/find.ml

```diff
@@ -130,3 +258,85 @@ let opt_label_in_sig s name =
     | [] -> None
   in
   inner s.Signature.items
+
+let find_in_sig sg f =
+  let rec inner = function
+    | Signature.Include i :: tl -> (
+        match inner i.Include.expansion_.items with
+        | Some _ as x -> x
+        | None -> inner tl )
+    | hd :: tl -> ( match f hd with Some _ as x -> x | None -> inner tl )
+    | [] -> None
+  in
+  inner sg.Signature.items
+
+let exception_in_sig s name =
+  find_in_sig s (function
+    | Signature.Exception (id, e) when Ident.Name.exception_ id = name -> Some e
+    | _ -> None)
+
+let extension_in_sig s name =
+  let rec inner = function
+    | ec :: _ when ec.Extension.Constructor.name = name -> Some ec
+    | _ :: tl -> inner tl
+    | [] -> None
+  in
+  find_in_sig s (function
+    | Signature.TypExt t -> inner t.Extension.constructors
+    | _ -> None)
+
+let label_parent_in_sig s name =
+  let module N = Ident.Name in
+  find_in_sig s (function
+    | Signature.Module (id, _, m) when N.module_ id = name ->
+        Some (`M (Component.Delayed.get m))
+    | ModuleType (id, mt) when N.module_type id = name ->
+        Some (`MT (Component.Delayed.get mt))
+    | Type (id, _, t) when N.type_ id = name ->
+        Some (`T (Component.Delayed.get t))
+    | Class (id, _, c) when N.class_ id = name ->
+        Some (`C c)
+    | ClassType (id, _, c) when N.class_type id = name
+      ->
+        Some (`CT c)
+    | _ -> None)
+
+let any_in_type_in_sig s name =
+  find_in_sig s (function
+    | Signature.Type (id, _, t) -> (
+        match any_in_type (Component.Delayed.get t) name with
+        | Some x -> Some (typename_of_typeid id, x)
+        | None -> None )
+    | _ -> None)
+
+let find_in_class_signature cs f =
+  let rec inner = function
+    | ClassSignature.Inherit ct_expr :: tl -> (
+        match inner_inherit ct_expr with Some _ as x -> x | None -> inner tl )
+    | it :: tl -> ( match f it with Some _ as x -> x | None -> inner tl )
+    | [] -> None
+  and inner_inherit = function
+    | Constr _ -> None
```

Comment:
Should we try to resolve inherited classes here ?

Reply:
Ugh, yes, we probably should! That's going to be a bit of a PITA.